Duplicated decorators during an edge case of named and unnamed decorators

m7pr opened the pull request:
Companion to https://github.com/insightsengineering/teal.modules.clinical/pull/1325

gogonzo commented on R/utils.R:
```diff
@@ -417,7 +417,9 @@ normalize_decorators <- function(decorators) {
-      list(default = decorators)
+      named <- which(names(decorators) != "")
+      unnamed <- setdiff(1:length(decorators), named)
+      c(list(default = decorators[unnamed]), lapply(decorators[named], list))
```
What happens if there are unnamed and named `default` in the same time ;]

llrs-roche replied:
Most (all?) modules uses

https://github.com/insightsengineering/teal.modules.general/blob/bfe87cce6173b8656d0c64b063ff1a80b4e70512/R/tm_a_pca.R#L189-L190

Testing the function with your example I see: 

```r
assert_decorators(normalize_decorators(decorators))
## Error: Assertion on 'decorators' failed: Must have names, but element 1 is empty.
```

But I agree we should make it more robust in case we start applying the function without an assert_decorators after it.

m7pr replied:
don't you miss available_decorators ?

m7pr replied:
For the mix of named+unnamed+named_with_default_name, I guess below function satifies edge cases. 
I commented out check for the `teal_transform_class` so we can work on regular lists for testing.

```r
normalize_decorators <- function(decorators) {
  #if (checkmate::test_list(decorators, "teal_transform_module")) {
    named <- which(names(decorators) != "")
    unnamed <- setdiff(seq_along(decorators), named)

    named_list <- lapply(decorators[named], list)

    if ("default" %in% names(named_list)) {
      # Merge unnamed elements into existing "default"
      named_list$default <- c(named_list$default, decorators[unnamed])
    } else {
      if (length(decorators[unnamed]) > 0) {
        # Create a new "default" entry
        named_list <- c(list(default = decorators[unnamed]), named_list)
      }
    }

    # Flatten the structure
    lapply(named_list, function(item){
      lapply(unlist(item), list)
    })
  # } else {
  #   return(decorators)
  # }
}
```

```r
decorators_1 <- list(
  list("unnamed1"),
  list("unnamed2"),
  default = list(
    list("default-1"),
    list("default-2")
  ),
  list("unnamed3"),
  named1 = list("named-value"),
  named2 = list(
    list("named2-value1"),
    list("named2-value2")
  )
)

decorators_2 <- list(
  list("unnamed1"),
  list("unnamed2")
)

decorators_3 <- list(
  named1 = list("named1"),
  named2 = list("named2")
)

decorators_4 <- list(
  named1 = list("named1"),
  default = list("default")
)

decorators_5 <- list(
  list("unnamed1"),
  default = list("default")
)



normalize_decorators(decorators_1) |> str()
normalize_decorators(decorators_2) |> str()
normalize_decorators(decorators_3) |> str()
normalize_decorators(decorators_4) |> str()
normalize_decorators(decorators_5) |> str()

```

```
> normalize_decorators(decorators_1) |> str()
List of 3
 $ default:List of 5
  ..$ :List of 1
  .. ..$ : chr "default-1"
  ..$ :List of 1
  .. ..$ : chr "default-2"
  ..$ :List of 1
  .. ..$ : chr "unnamed1"
  ..$ :List of 1
  .. ..$ : chr "unnamed2"
  ..$ :List of 1
  .. ..$ : chr "unnamed3"
 $ named1 :List of 1
  ..$ :List of 1
  .. ..$ : chr "named-value"
 $ named2 :List of 2
  ..$ :List of 1
  .. ..$ : chr "named2-value1"
  ..$ :List of 1
  .. ..$ : chr "named2-value2"
> normalize_decorators(decorators_2) |> str()
List of 1
 $ default:List of 2
  ..$ :List of 1
  .. ..$ : chr "unnamed1"
  ..$ :List of 1
  .. ..$ : chr "unnamed2"
> normalize_decorators(decorators_3) |> str()
List of 2
 $ named1:List of 1
  ..$ :List of 1
  .. ..$ : chr "named1"
 $ named2:List of 1
  ..$ :List of 1
  .. ..$ : chr "named2"
> normalize_decorators(decorators_4) |> str()
List of 2
 $ named1 :List of 1
  ..$ :List of 1
  .. ..$ : chr "named1"
 $ default:List of 1
  ..$ :List of 1
  .. ..$ : chr "default"
> normalize_decorators(decorators_5) |> str()
List of 1
 $ default:List of 2
  ..$ :List of 1
  .. ..$ : chr "default"
  ..$ :List of 1
  .. ..$ : chr "unnamed1"
```

m7pr replied:
Alrighty, for now I allowed only to pass either all named, or all unnamed decorators

<details><summary> Code </summary>

```r
pkgload::load_all(".")
data <- teal_data()
data <- within(data, {
  require(nestcolor)
  CO2 <- CO2
  factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
  CO2[factors] <- lapply(CO2[factors], as.character)
})

static_decorator <- teal_transform_module(
  label = "Static decorator",
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(), {
          plot <- plot +
            ggplot2::ggtitle("This is title") +
            ggplot2::xlab("x axis")
        })
      })
    })
  }
)

interactive_decorator <- teal_transform_module(
  label = "Interactive decorator",
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
               {
                 plot <- plot +
                   ggplot2::ggtitle("This is title") +
                   ggplot2::xlab(my_title)
               },
               my_title = input$x_axis_title
        )
      })
    })
  }
)

app <- init(
  data = data,
  modules = tm_g_bivariate(
    x = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "conc",
        fixed = FALSE
      )
    ),
    y = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "uptake",
        multiple = FALSE,
        fixed = FALSE
      )
    ),
    row_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Type",
        fixed = FALSE
      )
    ),
    col_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Treatment",
        fixed = FALSE
      )
    ),
    # decorators = list() # SHOULD BE OK
    # decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD BE OK
    # decorators = list(static_decorator, interactive_decorator) # SHOULD BE OK
    # decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD THROWN A WARNING
  )
)

shinyApp(app$ui, app$server)

```

</details>

llrs-roche reviewed (commented):
I can also reproduce the CI checks failures about the vignettes. 
Check the linter message and Dawid comment.

m7pr commented:
Alrighty, for now I allowed only to pass either all named, or all unnamed decorators

<details><summary> Code </summary>

```r
pkgload::load_all(".")
data <- teal_data()
data <- within(data, {
  require(nestcolor)
  CO2 <- CO2
  factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
  CO2[factors] <- lapply(CO2[factors], as.character)
})

static_decorator <- teal_transform_module(
  label = "Static decorator",
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(), {
          plot <- plot +
            ggplot2::ggtitle("This is title") +
            ggplot2::xlab("x axis")
        })
      })
    })
  }
)

interactive_decorator <- teal_transform_module(
  label = "Interactive decorator",
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
               {
                 plot <- plot +
                   ggplot2::ggtitle("This is title") +
                   ggplot2::xlab(my_title)
               },
               my_title = input$x_axis_title
        )
      })
    })
  }
)

app <- init(
  data = data,
  modules = tm_g_bivariate(
    x = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "conc",
        fixed = FALSE
      )
    ),
    y = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "uptake",
        multiple = FALSE,
        fixed = FALSE
      )
    ),
    row_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Type",
        fixed = FALSE
      )
    ),
    col_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Treatment",
        fixed = FALSE
      )
    ),
    # decorators = list() # SHOULD BE OK
    # decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD BE OK
    # decorators = list(static_decorator, interactive_decorator) # SHOULD BE OK
    # decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD THROWN A WARNING
  )
)

shinyApp(app$ui, app$server)

```

</details>

llrs-roche reviewed (changes requested):
I like the decision to only allow named or unnamed decorators. 

- `decorators = list()` 

  No problem ✅ 

- `decorators = list(plot = static_decorator, default = interactive_decorator) ` 

  I don't see the default interactive decorator having any effect. This makes sense but there is no message that the named one will be used instead. This confused me initially. Could you add a message/warning so that users know what is happening on that case? Or alternative patch the documentation? 
  
![image](https://github.com/user-attachments/assets/987ba4d7-8948-4a50-9032-64772ab10397)

  ⁉️ 

- `decorators = list(static_decorator, interactive_decorator)` 
  The interactive decorator worked well, but again no notification of what happened with the static_decorator. I would prefer to have some documentation or a clear message of what is happening here. :interrobang:

- `decorators = list(plot = static_decorator, interactive_decorator)`
  Throws an error (not a warning) ✅ 

- `decorators = list(plot = static_decorator, plot = interactive_decorator)`
  This example also throws the same error "All decorators should either be named or unnamed" but here the duplicated decorator are named so the message is not informative. Could the users get a more informative message? ❌

m7pr commented:
@llrs-roche funny things.

> I don't see the default interactive decorator having any effect. This makes sense but there is no message that the named one will be used instead. This confused me initially. Could you add a message/warning so that users know what is happening on that case? Or alternative patch the documentation?

Both of them are used, but both of them edit the same thing.

`static_decorator` is applied after `interactive_decorator` 
they both edit `ggplot2::xlab("x axis")`.

`interactive_decorator` uses input to edit `xlab` but then `static_decorator` overwrites with `ggplot2::xlab("x axis")`.

If you remove `ggplot2::xlab("x axis")` from `static_decorator`, you can see that changes from `interactive_decorator` are actually visible:

<img width="733" alt="image" src="https://github.com/user-attachments/assets/d2e077d5-1548-4b34-a778-b3d8c5cd2c95" />


<details><summary> Code </summary>

```r
pkgload::load_all(".")
data <- teal_data()
data <- within(data, {
  require(nestcolor)
  CO2 <- CO2
  factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
  CO2[factors] <- lapply(CO2[factors], as.character)
})

static_decorator <- teal_transform_module(
  label = "Static decorator",
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(), {
          plot <- plot +
            ggplot2::ggtitle("This is title")
        })
      })
    })
  }
)

interactive_decorator <- teal_transform_module(
  label = "Interactive decorator",
  ui = function(id) {
    ns <- NS(id)
    div(
      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
    )
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      reactive({
        req(data())
        within(data(),
               {
                 plot <- plot +
                   ggplot2::ggtitle("This is title") +
                   ggplot2::xlab(my_title)
               },
               my_title = input$x_axis_title
        )
      })
    })
  }
)

app <- init(
  data = data,
  modules = tm_g_bivariate(
    x = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "conc",
        fixed = FALSE
      )
    ),
    y = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "uptake",
        multiple = FALSE,
        fixed = FALSE
      )
    ),
    row_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Type",
        fixed = FALSE
      )
    ),
    col_facet = data_extract_spec(
      dataname = "CO2",
      select = select_spec(
        label = "Select variable:",
        choices = variable_choices(data[["CO2"]]),
        selected = "Treatment",
        fixed = FALSE
      )
    ),
    # decorators = list() # SHOULD BE OK
    # decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD BE OK
    # decorators = list(static_decorator, interactive_decorator) # SHOULD BE OK
    decorators = list(plot = static_decorator, default = interactive_decorator) # SHOULD THROWN A WARNING
  )
)

shinyApp(app$ui, app$server)

```

</details>

m7pr commented:
> decorators = list(plot = static_decorator, plot = interactive_decorator)
This example also throws the same error "All decorators should either be named or unnamed" but here the duplicated decorator are named so the message is not informative. Could the users get a more informative message? ❌

I think we can work on this. Thanks for double checking

m7pr commented:
Hey @llrs-roche when non-unique names are passed to decorators, function will throw a warning from `assert_decorators`. I allowed such case to pass through `normalize_decorators`. Check last [commit.](https://github.com/insightsengineering/teal.modules.general/pull/835/commits/67d76edac92ddd1b7615d0e25066a6ca86d0b14f)

Now user will see

```r
Error in tm_g_bivariate(x = data_extract_spec(dataname = "CO2", select = select_spec(label = "Select variable:",  : 
  Assertion on 'decorators' failed: Non-unique names in decorators.
Called from: mstop("Assertion on '%s' failed: %s.", var.name, res, call. = sys.call(-2L))
```

I will push the same to tmc

llrs-roche reviewed (approved):
Looks good and the corner case was handled elegantly. 👍
